### ISSUE TITLE
hooks for new authentication system

### ISSUE THREAD

**Issue by eduo** [craigrodway/classroombookings#30]
Hello.

I saw the LDAP authentication and was wondering what would be the template for adding a new type of authentication. I want to load the app as an iframe inside an existing one (I would prefer to integrate it better, but codeIgniter apps don't really lend themselves to this) and I would really like to reuse existing authentication.

To me this means on one side adding to index.php my authentication check (so it can't be loaded without the parent frame's authentication) and on another to call the user details and logged in status from my own.

I have been taking a loot at it but I can't figure out the current order of events that I would need to catch or how to create a new type of authentication that doesn't rely on a local user table at all (my user table is in a different database, and all information relevant to the user lives in the cookie).

Any help or guidance? What are the least intrusive modifications I could do to take over user authentication check and user tokens and ID.

**Comment by craigrodway** [comment]
Hi! There are a couple of things to unpick here but I'll try to answer as best I can!

First, the classroombookings users table is required, as it ties into a few different areas. Even if it isn't used directly for authenticating someone, the 'current logged-in user' needs to have an entry in the users table for things to work properly.

You didn't mention any details about the current authentication you want to re-use, but usually when I need to do something like this I will use a process called 'pre-authentication'.

The general idea with this method is as follows:

- Users will log in to your existing system first.
- In the existing system, you will generate a token of some sort - which would typically contain a user identifier, name/email, and timestamp. 
- This token is signed with a secret key that is known between the two systems. 
- The existing system presents the token to the second system.
- The second system checks for the presence of a token and verifies it using the secret key and compares timestamps.
- If it passes validation, it knows it is trustworthy - so it can inspect it for the user details and carry out any necessary actions.
- At this point you can then create the user record and/or log them in.

I would look at using a [JWT](https://jwt.io/) for this, as it's a good standard. I'd use the firebase/php-jwt package to deal with the validation/parsing - it's a great library.

In terms of where to do this, I would suggest using CodeIgniter's [hooks system](https://codeigniter.com/userguide3/general/hooks.html) and adding one or two methods to the `pre_controller` one to handle the process described above.

So you could, for example, generate a token and then set the iframe URL to `classroombookings.example.com/?auth_token=abc123zyz`. In the code you create for the hook, look for `$CI->input->get('auth_token')` and process it accordingly. Load and call the Users_model to check for existing users or create them, and then call the Userauth library to log them in (you can call the `log_in()` method with `$force=true` to log a user in without a password).

Official JWT authentication support is on the roadmap, but there are some other features on the top of the pile first.

Hope that's useful!

**Comment by eduo** [comment]
Thanks for the quick reply.

My other system's users table (and cookie) contains all the same data for the user as the users table would yield (sans password). The system is in the same server (not the same database) in a parent directory and I really wanted to remove the users table (since all user provisioning should happen elsewhere and keeping the user list synchronized really just adds another potential GDPR problem to my current pile :D ).

Authentication currently uses SAML ADFS for pre-authentication and then all is handled through the local user library of my main app (which keeps no passwords).

Essentially my system already does what you suggest: Authentication happens elsewhere and the system trusts the third party when serving each page that the user is who he says to be.

Once SAML was set-up, every new "module" I've added has just required a handful of calls at the beginning of index.php, which in turn check authorization and get the user data. 

I was hoping to do this as well here but the best I've got is a login page that is also authenticated in my main system :)

I realized I can't trace what the app is doing to decide the user is not authenticated so it doesn't present the login screen (my cookie contains data that can be mapped to all in the users table, except for the password which shouldn't be necessary).

My plan of supplanting the routine that decides a user is authenticated (and removing all functions that modify the users table) didn't work out :)

**Comment by eduo** [comment]
Hello. 

I managed to get this done by:
- Calling my pre-authorisation routine to only allow the page to load if the user is already authorised for the site.
- In userauth's init_user I used the UUID from my pre-auth and I get all necessary $user data. From that I build the $_SESSION and comment out the original $user function (CI->...->get_by_id

Other than this, the rest were aesthetic changes (removing header, fixing some widths and style assumptions so the page can load both standalone and in a frame, removing hardcoded pixels for better Smartphone display, etc).

All quick & dirty and possibly not worth of a pull request. But I'm leaving this here in case someone can find it useful.
